int64 enc1
int64 enc2